src/utils.py

```diff
@@ -73,3 +78,129 @@ def decorated(self, event):
+
+    Args:
+        length: number of characters to generate
+
+    Returns:
+        String of randomized letter+digit characters
+    """
+    uppercase_letters = string.ascii_uppercase
+    lowercase_letters = string.ascii_lowercase
+    digits = string.digits
+
+    all_characters = uppercase_letters + lowercase_letters + digits
+
+    characters = [
+        secrets.choice(uppercase_letters)
+        + secrets.choice(lowercase_letters)
+        + secrets.choice(digits)
+    ]
+    characters.extend(
+        secrets.choice(all_characters) for _ in range(length - 3)
+    )
+    secrets.SystemRandom().shuffle(characters)
+    password = "".join(characters)
+
+    return password
+
+
+def retry(max_retries=3, delay=2, backoff=2):
+    """Decorate function to retry executing upon failure.
+
+    Args:
+        max_retries: The maximum number of times to retry the decorated function.
+        delay: The initial delay (in seconds) before the first retry.
+        backoff: The factor by which the delay increases with each retry.
+
+    Returns:
+        decorator: A retry decorator function.
+    """
+
+    def decorator(func):
+        """Apply decorator function to the target function.
+
+        Args:
+            func: The function to decorate.
+
+        Returns:
+            wrapper: A decorated function that will be retried upon failure.
+        """
+
+        def wrapper(*args, **kwargs):
+            """Execute wrapper for the decorated function and handle retries.
+
+            Args:
+                args: Positional arguments passed to the decorated function.
+                kwargs: Keyword arguments passed to the decorated function.
+
+            Returns:
+                result: The result of the decorated function if successful.
+                None: If max_retries are reached without success, returns None.
+
+            Raises:
+                RangerServiceException: If max_retries are reached without success.
+            """
+            logger = logging.getLogger(__name__)
+
+            current_delay = delay  # Define current_delay before using it
+            for attempt in range(max_retries):
+                try:
+                    result = func(*args, **kwargs)
+                    return result
+                except Exception as e:
+                    if attempt < max_retries - 1:
+                        logger.warning(
+                            f"Request failed (attempt {attempt + 1}): {e}"
+                        )
+                        time.sleep(current_delay)
+                        current_delay *= backoff
+                    else:
+                        logger.exception("Max retries reached for request")
+                        raise RangerServiceException(
+                            "Max retries reached for request."
+                        ) from e
+            return None
+
+        return wrapper
+
+    return decorator
+
+
+def handle_service_error(func):
```

Comment:
This is not really handling anything. I guess you use this pattern often enough that you turned it into a decorator, which might have been a past suggestion of mine. Is that the reason?